pkg/dynamodb/dynamo_api.go

```diff
@@ -0,0 +1,419 @@
+}
+
+func (dynamoClient *DynamoClient) UpdateItem(ctx context.Context, inputRequest *UpdateItemParams, outputItem interface{}) error {
+	expr, err := makeUpdateExpression(inputRequest)
+
+	if err != nil {
+		return err
+	}
+
+	key, err := marshalMap(inputRequest.KeyCondition)
+
+	if err != nil {
+		return err
+	}
+
+	queryInput := &dynamodb.UpdateItemInput{
+		TableName:                 aws.String(dynamoClient.Table),
+		Key:                       key,
+		UpdateExpression:          expr.Update(),
+		ExpressionAttributeValues: expr.Values(),
+		ExpressionAttributeNames:  expr.Names(),
+		ReturnValues:              types.ReturnValueAllNew,
+		ConditionExpression:       expr.Condition(),
+	}
+
+	res, err := dynamoClient.updateItem(ctx, queryInput)
+
+	if err != nil {
+		return err
+	}
+
+	if outputItem != nil {
+		err = unmarshalMap(res.Attributes, outputItem)
+	}
+
+	return err
+}
+
+func makeUpdateExpression(inputRequest *UpdateItemParams) (expression.Expression, error) {
+	update := expression.UpdateBuilder{}
+	if inputRequest.ValuesToBeUpdated != nil {
+		inputRequest.ValuesToBeUpdated["lastUpdated"] = time.Now().Unix()
+	} else {
+		inputRequest.ValuesToBeUpdated = map[string]interface{}{
+			"lastUpdated": time.Now().Unix(),
+		}
+	}
+
+	for key, val := range inputRequest.ValuesToBeUpdated {
+		update = update.Set(
+			name(key),
+			value(val),
+		)
+	}
+
+	for _, val := range inputRequest.ValuesToBeRemoved {
+		update = update.Remove(
+			name(val),
+		)
+	}
+
+	for _, val := range inputRequest.ValuesToBeIncremented {
+		update.Set(
+			name(val),
+			plus(val, 1),
+		)
+	}
+
+	for key, val := range inputRequest.ValuesToBeAppended {
+		update = update.Set(
+			name(key),
+			listAppend(key, val),
+		)
+	}
+
+	conditions, err := setFilterConditionBuilder(inputRequest.ConditionalParams)
+	if err != nil {
+		return expression.Expression{}, err
+	}
+
+	expressionBuilder := expression.NewBuilder()
+	expressionBuilder = expressionBuilder.WithUpdate(update)
+
+	if conditions.IsSet() {
+		expressionBuilder = expressionBuilder.WithCondition(conditions)
+	}
+
+	expr, err := expressionBuilder.Build()
+
+	if err != nil {
+		return expr, err
+	}
+
+	return expr, nil
+}
+
+func (dynamoClient *DynamoClient) DeleteItem(ctx context.Context, inputRequest DeleteItemParams) error {
+	row, err := marshalMap(inputRequest.KeyCondition)
+
+	if err != nil {
+		return err
+	}
+
+	input := &dynamodb.DeleteItemInput{
+		Key:       row,
+		TableName: aws.String(dynamoClient.Table),
+	}
+
+	_, err = dynamoClient.deleteItem(ctx, input)
+
+	return err
+}
+
+func (dynamoClient *DynamoClient) QueryItems(
+	ctx context.Context,
+	queryItemsParams *QueryItemsParams,
+	outputItem interface{},
+) ([]map[string]types.AttributeValue, int, error) {
+	projectionBuilder := setProjectionBuilder(queryItemsParams.ProjectionKeys)
+
+	conditionBuilder, err := setConditionBuilder(queryItemsParams.PartitionKey, queryItemsParams.SortKey)
+	if err != nil {
+		return nil, 0, err
```

Comment:
**Suggestion:** In the `QueryItems` method, consider checking if `queryItemsParams` is nil before accessing its fields to prevent potential nil pointer dereference. [possible bug, importance: 8]
```suggestion
 if queryItemsParams == nil {
     return nil, 0, errors.New("queryItemsParams cannot be nil")
 }
 projectionBuilder := setProjectionBuilder(queryItemsParams.ProjectionKeys)

 conditionBuilder, err := setConditionBuilder(queryItemsParams.PartitionKey, queryItemsParams.SortKey)
 if err != nil {
     return nil, 0, err
 }
```